src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs

```diff
@@ -74,7 +74,13 @@ public virtual async Task StopAsync(CancellationToken cancellationToken)
             finally
             {
                 // Wait until the task completes or the stop token triggers
-                await Task.WhenAny(_executeTask, Task.Delay(Timeout.Infinite, cancellationToken)).ConfigureAwait(false);
+#if NETCOREAPP
+                await _executeTask.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
Test are failing off of line 78. Here's a [stack trace](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-74789-merge-50e1925b33ea4871a5/Microsoft.Extensions.Hosting.Unit.Tests/1/console.50686ade.log?helixlogtype=result) of some of the fails.

Reply:
We were never actually `await`ing the `_executeTask` before. Before we were just `await`ing `Task.WhenAny` which successfully returns when either of those tasks completes (if that underlying Task is successful or not).

Basically we need to ignore that the task is getting cancelled. The first way I can think of to do that is to:

```C#
try
{
    await _executeTask.WaitAsync(cancellationToken).ConfigureAwait(false);
}
catch (OperationCanceledException)
{
    // ignore cancellation exceptions since we are stopping the service and the cancellation is expected
}
```

But maybe there is a cleaner way to have have exception processing in the "happy path".

Another option is to not use the new API and instead unconditionally use the `TaskCompletionSource` approach even on NETCOREAPP.

Reply:
> We were never actually awaiting the _executeTask before.

The intention was to await _executeTask, though; the original code had a bug (#36380).

Reply:
It looks like the behavior of BackgroundService handling [has changed significantly](https://github.com/dotnet/runtime/blob/b13e6ef8cba3a1be054e2d3fd2f431b2b14dbf34/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L73-L76) from what it was when I opened #36380.  @Nick-Stanton, you should keep in mind that my original suggestion was for .NET Core 3.0, and the behavior has changed since then.  My suggestion is no longer valid with the current behavior.  Line 82 in your current version is also erroneous with the current behavior: it shouldn't await twice; that was only valid when BackgroundService [didn't expose _executeTask](https://github.com/dotnet/extensions/blob/aa7fa91cfc8f6ff078b020a428bcad71ae7a32ab/src/Hosting/Abstractions/src/BackgroundService.cs#L15) (the ExecuteTask property didn't exist).

Reply:
Looking at the !NETCOREAPP behavior, this is actually a bit of a mess: if _executeTask is returned, it must not be awaited, but if tcs.Task is returned, it must be awaited.  See line 82 below.